src/network/hobbits/hobbitsConnectHandler.ts

```diff
@@ -0,0 +1,266 @@
+    if (this.peers.has(peerId)) {
+      return;
+    }
+    if (this.wipDials.has(peerId)) {
+      return;
+    }
+    this.wipDials.add(peerId);
+    try {
+      const peer = this.addPeer(hobbitsUri);
+      peer.connect();
+    } catch (e) {
+      this.logger.error(e.stack);
+    }
+    this.wipDials.delete(peerId);
+  }
+
+  public async sendRequest<T extends ResponseBody>(hobbitsUri: HobbitsUri, type: ProtocolType, method: Method, body: RequestBody): Promise<T> {
+    const peerId = hobbitsUri.toUri();
+    const peer = this.peers.get(peerId);
+    if (!peer) {
+      throw new Error(`Peer does not exist: ${peerId}`);
+    }
+    let encodedRequest;
+    const id = randomRequestId();
+    if (type == ProtocolType.RPC) {
+      this.requests[id] = method;
+      encodedRequest = encodeRequest(id, method, body);
+    }
+
+    const encodedMessage = encodeMessage(type, encodedRequest);
+    peer.write(encodedMessage);
+    return await this.getResponse(id) as T;
+  }
+
+  private async getResponse(id: RequestId): Promise<ResponseBody> {
+    return new Promise<ResponseBody>((resolve, reject) => {
+      const timeout = setTimeout(() => {
+        this.removeAllListeners(id+'');
+        const method = this.requests[id];
+        delete this.requests[id];
+        reject(new Error(`request timeout, method ${method}, id ${id}`));
+      }, this.requestTimeout);
+      this.once(`response ${id}`, (err, data) => {
+        clearTimeout(timeout);
+        delete this.requests[id];
+        if (err) {
+          reject(err);
+        } else {
+          resolve(data);
+        }
+      });
+    });
+  }
+
+  // public sendResponse(id: RequestId, responseCode: number, result: ResponseBody): void {
+  //   const request = this.responses[id];
+  //   if (!request) {
+  //     throw new Error('No request found');
+  //   }
+  //   const {peer, method} = request;
+  //   const encodedResponse = encodeResponse(id, method, responseCode, result);
+  //   delete this.responses[id];
+  //   peer.write(encodedResponse);
+  // }
+
+  public onRequestResponse(peer: Peer, data: Buffer): void {
+    if(!sanityCheckData(data)) {
+      // bad data
+      return;
+    }
+    // Changed response
+    let decodedBody, request;
+    let decodedMessage = decodeMessage(data);
+    switch (decodedMessage.protocol) {
+      case ProtocolType.RPC:
+        request = deserialize(decodedMessage.payload, WireRequest);
+        decodedBody = decodeRequestBody(request.methodId, request.body);
+        break;
+      case ProtocolType.GOSSIP:
+        break;
+      case ProtocolType.PING:
+        break;
+    }
+
+    const id = request.id;
+    const method = this.requests[id];
+    if (method === undefined) { // incoming request
+      try {
+        this.responses[id] = {peer, method: request.method};
+        this.emit("request", peer.hobbitsUri, request.method, id, decodedBody);
+      } catch (e) {
+        this.logger.warn('unable to decode request', e.message);
```

Comment:
I don't think any of these method will throw, rather wrap decodeMessage in try catch and/or validate parsed data